kubelet: group all container-runtime-specific flags/options into a separate struct

yujuhong opened the pull request:
They don't belong in the KubeletConfig.

This addresses #43253

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/44061)
<!-- Reviewable:end -->

yujuhong commented:
/cc @Random-Liu

supereagle commented on cmd/kubelet/app/options/options.go:
```diff
@@ -80,6 +80,7 @@ type KubeletFlags struct {
 type KubeletServer struct {
 	KubeletFlags
 	componentconfig.KubeletConfiguration
+	DockerOptions *DockerOptions
```
Why not take `DockerOptions` as an embedded field of `KubeletServer` like `KubeletFlags`? Then the fields of `DockerOptions` can be directly accessed.

yujuhong replied:
This is intentional. `DockerOptions` may be nil in the future.

supereagle commented on cmd/kubelet/app/server.go:
```diff
@@ -156,7 +156,7 @@ func UnsecuredKubeletDeps(s *options.KubeletServer) (*kubelet.KubeletDeps, error
 		ExternalKubeClient: nil,
 		Mounter:            mounter,
-		NetworkPlugins:     ProbeNetworkPlugins(s.NetworkPluginDir, s.CNIConfDir, s.CNIBinDir),
+		NetworkPlugins:     ProbeNetworkPlugins(s.DockerOptions.NetworkPluginDir, s.DockerOptions.CNIConfDir, s.DockerOptions.CNIBinDir),
```
If take `DockerOptions` as an embedded field of `KubeletServer`, these and other changes are not needed, as the fields of DockerOptions such as `s.DockerOptions.NetworkPluginDir` can be directly accessed by `s.NetworkPluginDir` like before.

yujuhong commented:
Unfortunately some of the flags are shared by rktnetes, which complicates things. I am going to go through the list of options again. Tagging this as WIP.

xiangpengzhao commented on cmd/kubelet/app/options/docker.go:
```diff
@@ -0,0 +1,81 @@
+	DockerEndpoint string
+	// networkPluginName is the name of the network plugin to be invoked for
+	// various events in kubelet/pod lifecycle
+	NetworkPluginName string
```
Why do we take these network flags as docker options?

yujuhong replied:
In the new CRI world, network plugins are container runtime details. kubelet is not aware of them.

The tricky part is that these are still consumed by rkt since they haven't switched to CRI. That's why I marked this PR as WIP.

xiangpengzhao replied:
Thanks for explanation!

xiangpengzhao commented on cmd/kubelet/app/options/options.go:
```diff
@@ -248,8 +241,6 @@ func (c *kubeletConfiguration) addFlags(fs *pflag.FlagSet) {
 	fs.Int32Var(&c.MaxPods, "max-pods", c.MaxPods, "Number of Pods that can run on this Kubelet.")
-	// TODO(#40229): Remove the docker-exec-handler flag.
-	fs.StringVar(&c.DockerExecHandlerName, "docker-exec-handler", c.DockerExecHandlerName, "Handler to use when executing a command in a container. Valid values are 'native' and 'nsenter'. Defaults to 'native'.")
 	fs.MarkDeprecated("docker-exec-handler", "this flag will be removed and only the 'native' handler will be supported in the future.")
```
delete this line?

yujuhong replied:
As the issue states, we can't really remove this yet.

mtaufen commented:
Should `--enable-custom-metrics` also get demoted to flag-only?

mtaufen commented:
And what about fields like:
```
  // dockerEndpoint is the path to the docker endpoint to communicate with.
  DockerEndpoint string `json:"dockerEndpoint"`

  // dockerExecHandlerName is the handler to use when executing a command
  // in a container. Valid values are 'native' and 'nsenter'. Defaults to
  // 'native'.
  DockerExecHandlerName string `json:"dockerExecHandlerName"`

  // rktPath is the  path of rkt binary. Leave empty to use the first rkt in
  // $PATH.
  RktPath string `json:"rktPath"`

  // rktApiEndpoint is the endpoint of the rkt API service to communicate with.
  RktAPIEndpoint string `json:"rktAPIEndpoint"`

  // rktStage1Image is the image to use as stage1. Local paths and
  // http/https URLs are supported.
  RktStage1Image string `json:"rktStage1Image"
```

This stuff seems pretty runtime-specific and probably shouldn't be fed through the Kubelet.

yujuhong commented:
> And what about fields like:
> 
>   // dockerEndpoint is the path to the docker endpoint to communicate with.
>   DockerEndpoint string `json:"dockerEndpoint"`
> 
>   // dockerExecHandlerName is the handler to use when executing a command
>   // in a container. Valid values are 'native' and 'nsenter'. Defaults to
>   // 'native'.
>   DockerExecHandlerName string `json:"dockerExecHandlerName"`
> 
>   // rktPath is the  path of rkt binary. Leave empty to use the first rkt in
>   // $PATH.
>   RktPath string `json:"rktPath"`
> 
>   // rktApiEndpoint is the endpoint of the rkt API service to communicate with.
>   RktAPIEndpoint string `json:"rktAPIEndpoint"`
> 
>   // rktStage1Image is the image to use as stage1. Local paths and
>   // http/https URLs are supported.
>   RktStage1Image string `json:"rktStage1Image"
> This stuff seems pretty runtime-specific and probably shouldn't be fed through the Kubelet.

The docker endpoint and exec handler are already included in the `DockerOptions` in this PR :-|
Not sure about the rkt fields yet. I am not too keen on creating a generic "container runtime options" since rkt and docker are integrated in very different ways (non-CRI vs. CRI) today.

mtaufen commented:
Please ping me when this is ready for re-review.

mtaufen reviewed (commented):
We should also use this PR to mark deprecated any runtime-related flags that are planned for removal from the Kubelet.

yujuhong committed:
Group container-runtime-specific flags/options together

Do not store them in kubelet's configuration. Eventually, we would like
to deprecate all these flags as they should not be part of kubelet.

yujuhong commented:
@mtaufen I changed the PR to include all runtime-specific flags. It's not complete though. I think it's ok as a start. Let me know what you think.

mtaufen commented on cmd/kubelet/app/options/options.go:
```diff
@@ -87,6 +83,7 @@ type KubeletFlags struct {
 // a kubelet. These can either be set via command line or directly.
 type KubeletServer struct {
 	KubeletFlags
+	ContainerRuntimeOptions
```
Not in this PR, but very soon we'll have to actually name these sub-objects. It's getting very confusing what comes from where in KubeletServer.

mtaufen replied:
Also, would it make sense for these to be subordinate to the KubeletFlags object, rather than directly on the KubeletServer, given that they all come in as flags-only?

yujuhong replied:
I named it in the original PR but changed it to be consistent with other fields in the struct. Will switch it back.

> Also, would it make sense for these to be subordinate to the KubeletFlags object, rather than directly on the KubeletServer, given that they all come in as flags-only?

Make sense. We still need to pass the container runtime options as a whole to kubelet, but I think that's fine. Done.

mtaufen replied:
Yeah, that's fine. Thanks!

mtaufen commented on pkg/kubemark/hollow_kubelet.go:
```diff
@@ -172,6 +175,6 @@ func GetHollowKubeletConfig(
 	//                have actually wanted the default).
 	//                The default will be present in the KubeletConfiguration contstructed above.
```
You can delete this TODO as well. It's old and irrelevant these days.

yujuhong replied:
Done.

mtaufen replied:
Thanks!

mtaufen reviewed (commented):
Nice work, I had a few comments.

yujuhong commented:
> We should also use this PR to mark deprecated any runtime-related flags that are planned for removal from the Kubelet.

I'll continue evaluate other options to see what can be moved to this new struct, but I think getting this in would help prevent people from adding more fields to KubeletConfig. 
As for deprecating the flags, it's to early to do that since we don't have replacement yet.

mtaufen commented:
Yeah, SGTM

mtaufen commented:
/lgtm 
/approve

mtaufen commented:
@wojtek-t for approval

yujuhong commented:
Squashed the commits to make review easier.

yujuhong commented:
@k8s-bot unit test this
@k8s-bot node-e2e test this
@k8s-bot e2e-gce-etcd3 test this
@k8s-bot e2e-kops-aws test this

yujuhong commented:
@k8s-bot node-e2e test this
@k8s-bot e2e-gce-etcd3 test this
@k8s-bot e2e-kops-aws test this

yujuhong commented:
/cc @gmarek for approval.

yujuhong commented:
@k8s-bot pull-kubernetes-unit test this

yujuhong commented:
@k8s-bot pull-kubernetes-unit test this

yujuhong commented:
@k8s-bot pull-kubernetes-e2e-kops-aws test this

I've been pasting the wrong one :-|

shyamjvs commented:
/approve

yujuhong commented:
Assigned to @shyamjvs. I found a new victim to approve the kubemark changes before I have to rebase again :-)

yujuhong commented:
Rebased and re-applying the lgtm & approved labels based on https://github.com/kubernetes/kubernetes/pull/44061#issuecomment-304126014 and https://github.com/kubernetes/kubernetes/pull/44061#issuecomment-304413750
